base/libgit2/types.jl

```diff
@@ -754,50 +735,31 @@ function securezero!(cred::SSHCredentials)
     securezero!(cred.pass)
     securezero!(cred.pubkey)
     securezero!(cred.prvkey)
+    cred.count = 0
     return cred
 end
 
 "Credentials that support caching"
 type CachedCredentials <: AbstractCredentials
-    cred::Dict{String,SSHCredentials}
+    cred::Dict{String,AbstractCredentials}
     count::Int            # authentication failure protection count
-    CachedCredentials() = new(Dict{String,SSHCredentials}(),3)
-end
-"Returns specific credential parameter value: first index is a credential
-parameter name, second index is a host name (with schema)"
-function Base.getindex(p::CachedCredentials, keys...)
-    length(keys) != 2 && return nothing
-    key, host = keys
-    if haskey(p.cred, host)
-        creds = p.cred[host]
-        if isdefined(creds,key)
-            kval = getfield(creds, key)
-            !isempty(kval) && return kval
-        end
-    end
-    return nothing
-end
-"Sets specific credential parameter value: first index is a credential
-parameter name, second index is a host name (with schema)"
-function Base.setindex!(p::CachedCredentials, val, keys...)
-    length(keys) != 2 && return nothing
-    key, host = keys
-    if !haskey(p.cred, host)
-        p.cred[host] = SSHCredentials()
-    end
-    creds = p.cred[host]
-    @assert isdefined(creds,key)
-    setfield!(creds, key, val)
-    return p
+    CachedCredentials() = new(Dict{String,AbstractCredentials}(),3)
 end
+
 "Checks if credentials were used or failed authentication, see `LibGit2.credentials_callback`"
-function checkused!(p::CachedCredentials)
+function checkused!(p::Union{UserPasswordCredentials, SSHCredentials})
     p.count <= 0 && return true
     p.count -= 1
     return false
 end
-"Resets authentication failure protection count"
-reset!(p::CachedCredentials, cnt::Int=3) = (p.count = cnt)
+reset!(p::Union{UserPasswordCredentials, SSHCredentials}, cnt::Int=3) = (p.count = cnt)
+reset!(p::CachedCredentials) = foreach(reset!, values(p.cred))
```

Comment:
Ah, yes. One problem could be that we do not know if the callback is called first time or after failed authentication when using cached credentials. This only could happen if following

1. the user entered incorrect credentials
2. credentials were cached
3. a connection error happened during authentication, resulting in the operation failure
4. next operation started with cached incorrect credentials
5. infinite loop authentication loop with incorrect credentials !!!

More to that, the behavior is protocol dependent:

- `http`: Only one credential type can be used: `GIT_CREDTYPE_USERPASS_PLAINTEXT`. The credentials callback function is called again when authentication was unsuccessful. 
- `ssh`: Multiple credential types can be used: `GIT_CREDTYPE_USERPASS_PLAINTEXT`, `GIT_CREDTYPE_SSH_KEY`, `GIT_CREDTYPE_SSH_CUSTOM`, `GIT_CREDTYPE_SSH_INTERACTIVE`. When credentials callback function is called again when authentication was unsuccessful, *but it is not possible to find which credential type was used*.